worker/sshserver/server.go

```diff
@@ -0,0 +1,124 @@
+	"fmt"
+	"io"
+	"net"
+
+	"github.com/gliderlabs/ssh"
+	"github.com/juju/errors"
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/juju/juju/state"
+)
+
+// NewSSHServer returns an embedded SSH server. This server does a few things,
+// reuqests that come in must go through the jump server. The jump server will
+// pipe the connection and pass it into an in-memory instance of another SSH server.
+//
+// This second SSH server (seen within directTCPIPHandlerClosure) will handle the
+// the termination of the SSH connections, note, it is not listening on any ports
+// because we are passing the piped connection to it, essentially allowing the following
+// to work (despite only having one server listening):
+// - `ssh -J controller:2223 ubuntu@app.controller.model`
+//
+// TODO(ale8k): Word this comment better later explaining why the host routing works.
+func NewSSHServer(sp *state.StatePool, jumpHostKey string) (*ssh.Server, error) {
+	jumpHostSigner, err := gossh.ParsePrivateKey([]byte(jumpHostKey))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	server := &ssh.Server{
+		Addr: ":2223",
+		PublicKeyHandler: func(ctx ssh.Context, key ssh.PublicKey) bool {
+			return true
+		},
+		PasswordHandler: func(ctx ssh.Context, password string) bool {
+			return true
+		},
+		ChannelHandlers: map[string]ssh.ChannelHandler{
+			"direct-tcpip": directTCPIPHandlerClosure(sp),
+		},
+	}
+
+	server.AddHostKey(jumpHostSigner)
+
+	return server, nil
+}
+
+// directTCPIPHandlerClosure is a closure that returns a direct-tcpip handler passing in state
+// to check the permissions the user has for the model containing this unit.
+func directTCPIPHandlerClosure(sp *state.StatePool) ssh.ChannelHandler {
+	return func(srv *ssh.Server, conn *gossh.ServerConn, newChan gossh.NewChannel, ctx ssh.Context) {
+		d := struct {
+			DestAddr string
+			DestPort uint32
+			SrcAddr  string
+			SrcPort  uint32
+		}{}
+
+		if err := gossh.Unmarshal(newChan.ExtraData(), &d); err != nil {
+			newChan.Reject(gossh.ConnectionFailed, "Failed to parse channel data")
+			return
+		}
+
+		ch, reqs, err := newChan.Accept()
+		if err != nil {
+			return
+		}
+
+		go gossh.DiscardRequests(reqs)
+
+		jumpServerPipe, terminatingServerPipe := net.Pipe()
+
+		go func() {
+			defer ch.Close()
+			defer jumpServerPipe.Close()
+			defer terminatingServerPipe.Close()
+			io.Copy(ch, jumpServerPipe)
+		}()
+		go func() {
+			defer ch.Close()
+			defer jumpServerPipe.Close()
+			defer terminatingServerPipe.Close()
+			io.Copy(jumpServerPipe, ch)
+		}()
+
+		forwardHandler := &ssh.ForwardedTCPHandler{}
+		server := &ssh.Server{
+			PublicKeyHandler: func(ctx ssh.Context, key ssh.PublicKey) bool {
+				return true
+			},
+			LocalPortForwardingCallback: ssh.LocalPortForwardingCallback(func(ctx ssh.Context, dhost string, dport uint32) bool {
+				return true
+			}),
+			ReversePortForwardingCallback: ssh.ReversePortForwardingCallback(func(ctx ssh.Context, host string, port uint32) bool {
```

Comment:
The spec states we won't allow reverse port forwarding, so you can always return False here.

Reply:
I'll doc it too